### ISSUE TITLE
Listener X11 crashes and logs me out

### ISSUE THREAD

**Issue by rbreaves** [moses-palmer/pynput#269]
OS: Ubuntu Mate 20.04
Python: 3.8

I am not sure what I am doing wrong here, but hopefully someone can help point me in the right direction.. I've assumed that the listener must not be closing properly but when I've read it back out it appears like it closed from what I can tell.. I am just at a loss.

I even added a 3 second countdown timer, and sometimes it seems like it helps.. but then other times not really. I just don't know what is causing the instability.

```
#!/usr/bin/env python3
# pip3 install pynput
from pynput.keyboard import Key, Listener
import sys, subprocess, time
# , keyboard
delay=3
# listener = None
# keyPressed = None

class color:
   PURPLE = '\033[95m'
   CYAN = '\033[96m'
   DARKCYAN = '\033[36m'
   BLUE = '\033[94m'
   GREEN = '\033[92m'
   YELLOW = '\033[93m'
   RED = '\033[91m'
   BOLD = '\033[1m'
   UNDERLINE = '\033[4m'
   END = '\033[0m'

def countdown(secs):
    for i in range(0,secs):
        # print(secs-i,end="")
        print(secs-i, end="\r", flush=True)
        time.sleep(1)

def on_press(key):
    print('{0} pressed'.format(key))

def on_release(key):
    # print('{0} release'.format(key))
    # print()
    if key == Key.esc:
        # Stop listener
        return False

def yn_choice(message, default='y'):
    choices = 'Y/n' if default.lower() in ('y', 'yes') else 'y/N'
    choice = input("%s (%s) " % (message, choices))
    values = ('y', 'yes', '') if choices == 'Y/n' else ('y', 'yes')
    return choice.strip().lower() in values

modifier_keys = {
    "primary":"",
    "secondary":"",
    "rprimary":"",
    "rsecondary":""
}

def set_key(key):
    global modifiery_keys
    print("\nWhich key would you like to set?\n")

    while True:
        try:
            keytype = int(input(
                "1) Ctrl\n" +
                "2) Alt\n" +
                "3) Super/Win/Cmd/Chrome search key\n"))
            if keytype < 4 and keytype > 0:
                break
        except:
            print("That's not a valid option!")
    print("")
    if keytype == 1:
        modifier_keys[key] = "Ctrl"
    elif keytype == 2:
        modifier_keys[key] = "Alt"
    elif keytype == 3:
        modifier_keys[key] = "Cmd"

def is_primary(key):
    global modifiery_keys
    if not (str(key).replace("Key.", "").title() == "Enter" or str(key).replace("Key.", "").title() == "Esc"):
        print(str(key).replace("Key.", "").title() + " will be remapped to Ctrl, the Cmd ⌘  key position.")
        # countdown(3)
        modifier_keys["primary"] = str(key).replace("Key.", "").title()
        return False
    elif str(key).replace("Key.", "").title() == "Esc":
        modifier_keys["primary"] = "Esc"
        # countdown(3)
        return False
    else:
        return True

def is_secondary(key):
    global modifiery_keys
    if not (str(key).replace("Key.", "").title() == "Enter" or str(key).replace("Key.", "").title() == "Esc"):
        print(str(key).replace("Key.", "").title() + " will be remapped to Alt, the Option ⌥  key position.")
        # countdown(3)
        modifier_keys["secondary"] = str(key).replace("Key.", "").title()
        return False
    elif str(key).replace("Key.", "").title() == "Esc":
        modifier_keys["secondary"] = "Esc"
        # countdown(3)
        return False
    else:
        return True

def is_rprimary(key):
    global modifiery_keys
    if not (str(key).replace("Key.", "").title() == "Enter" or str(key).replace("Key.", "").title() == "Esc"):
        print(str(key).replace("Key.", "").title() + " will be remapped to Ctrl, the Cmd ⌘  key position.")
        # countdown(3)
        modifier_keys["rprimary"] = str(key).replace("Key.", "").title()
        return False
    elif str(key).replace("Key.", "").title() == "Esc":
        modifier_keys["rprimary"] = "Esc"
        # countdown(3)
        return False
    else:
        return True

def is_rsecondary(key):
    global modifiery_keys
    if not (str(key).replace("Key.", "").title() == "Enter" or str(key).replace("Key.", "").title() == "Esc"):
        print(str(key).replace("Key.", "").title() + " will be remapped to Alt, the Option ⌥  key position.")
        # countdown(3)
        modifier_keys["rsecondary"] = str(key).replace("Key.", "").title()
        return False
    elif str(key).replace("Key.", "").title() == "Esc":
        modifier_keys["rsecondary"] = "Esc"
        # countdown(3)
        return False
    else:
        return True


print(color.UNDERLINE + color.YELLOW + "\n\nPlease ignore the FN key." + color.END + " FN cannot be remapped by software, some Thinkpads can swap it with Ctrl in the BIOS.\n")
input("Press Enter to continue...\n\n")
print(chr(27) + "[2J")

while True:
    global modifiery_keys
    
    
    print(color.UNDERLINE + color.YELLOW + "\n\nPress the 1st key Left of the spacebar" + color.END + " (Press Esc to set manaully)\n")
    print("    👇 \n")
    print(" □ □ ▣ ░░░░░░░\n")
    with Listener(
        on_release=is_primary,suppress=True) as listener:
            try:
                listener.join()
            except MyException as e:
                print('{0} was pressed'.format(e.args[0]))

    if modifier_keys["primary"] != "Esc":
        choice = yn_choice("Is this correct?")
        if(choice):
            print("Left Physical " + modifier_keys["primary"] + " = Ctrl/Cmd ⌘")
            # listener.stop()
            break
    else:
        set_key("primary")
        print("Left Physical " + modifier_keys["primary"] + " = Ctrl/Cmd ⌘\n")
        # listener.stop()
        input("Press Enter to continue...\n\n")
        break

print(chr(27) + "[2J")

print(listener)

while True:
    global modifiery_keys
    print(color.UNDERLINE + color.YELLOW + "\n\nPress the 2nd key Left of the spacebar" + color.END + " (Press Esc to set manaully)\n")
    print("  👇\n")
    print(" □ ▣ □ ░░░░░░░\n")
    
    with Listener(
        on_release=is_secondary,suppress=True) as listener:
            try:
                listener.join()
            except MyException as e:
                print('{0} was pressed'.format(e.args[0]))

    if modifier_keys["secondary"] != "Esc":
        choice = yn_choice("Is this correct?")
        if(choice):
            # listener.stop()
            # print("Left Physical " + modifier_keys["secondary"] + " = Alt/Option ⌥")
            break
    else:
        set_key("secondary")
        print("Left Physical " + modifier_keys["secondary"] + " = Alt/Option ⌥\n")
        # listener.stop()
        input("Press Enter to continue...\n\n")
        break

print(chr(27) + "[2J")

while True:
    global modifiery_keys
    print(color.UNDERLINE + color.YELLOW + "\n\nPress the 1st key Right of the spacebar" + color.END + " (Press Esc to set manaully)\n")
    print("        👇 \n")
    print(" ░░░░░░░ ▣ □ \n")
    
    with Listener(
        on_release=is_rprimary,suppress=True) as listener:
            try:
                listener.join()
            except MyException as e:
                print('{0} was pressed'.format(e.args[0]))

    if modifier_keys["rprimary"] != "Esc":
        choice = yn_choice("Is this correct?")
        if(choice):
            # listener.stop()
            # print("Right Physical " + modifier_keys["rprimary"] + " = Ctrl/Cmd ⌘")
            break
    else:
        set_key("rprimary")
        print("Right Physical " + modifier_keys["rprimary"] + " = Ctrl/Cmd ⌘\n")
        # listener.stop()
        input("Press Enter to continue...\n\n")
        break

print(chr(27) + "[2J")

while True:
    print(color.UNDERLINE + color.YELLOW + "\n\nPress the 2nd key Right of the spacebar" + color.END + " (Press Esc to set manaully)\n")
    print("          👇\n")
    print(" ░░░░░░░ □ ▣ \n")
    

    
    with Listener(
        on_release=is_rsecondary,suppress=True) as listener:
            try:
                listener.join()
            except MyException as e:
                print('{0} was pressed'.format(e.args[0]))

    if modifier_keys["rsecondary"] != "Esc":
        choice = yn_choice("Is this correct?")
        if(choice):
            # listener.stop()
            # print("Right Physical " + modifier_keys["rsecondary"] + " = Alt/Option ⌥")
            break
    else:
        set_key("rsecondary")
        print("Right Physical " + modifier_keys["rsecondary"] + " = Alt/Option ⌥\n")
        # listener.stop()
        input("Press Enter to continue...\n\n")
        break

print(chr(27) + "[2J")

if modifier_keys["secondary"] != "Ctrl":
    print(color.UNDERLINE + color.YELLOW + "GUI Usage (Physical Ctrl key)\n"+ color.END)
    print("Ctrl key will be mapped to Super. (Search key on chromebooks)\n")
    print("👇\n")
    print(" ▣ □ □ ░░░░░░░\n")

    print("Note: Super may still activate Ctrl based shortcuts\n")
    print("at times depending on application or system level shortcut.\n")
    print("This will only be done to align shortcuts to their expected functionality.\n")

    input("Press Enter to continue...\n\n")
    # print(chr(27) + "[2J")

    print(color.UNDERLINE + color.YELLOW + "Terminal Usage" + color.END + "\n")
    print("Ctrl key will be the Ctrl key.\n")
    print("👇\n")
    print(" ▣ □ □ ░░░░░░░\n\n")
    print("The Cmd ⌘  key position during terminal usage will usually be Ctrl+Shift.\n")
    print("    👇 \n")
    print(" □ □ ▣ ░░░░░░░\n\n")
else:
    print("Chromebook detected.")

    print(color.UNDERLINE + color.YELLOW + "Terminal Usage" + color.END + "\n")
    print(" □ capslock/search key = Alt\n")
    print(" shift\n")
    print(" ▣ □ ░░░░░░░")
    print("☝️\n")

    input("Press Enter to continue...\n\n")
```

**Comment by mohammedyunus009** [comment]
OMG . Even i facing the same issue here (ubuntu 16). My env details .

Name: pynput
Version: 1.6.8
Summary: Monitor and control user input devices
Home-page: https://github.com/moses-palmer/pynput
Author: Moses Palmér
Author-email: moses.palmer@gmail.com
License: LGPLv3
Location: /home/*********/.conda/envs/snakes/lib/python3.7/site-packages
Requires: python-xlib, six
Required-by:

**Comment by rbreaves** [comment]
Hmmmm wonder if I should try earlier versions..

**Comment by moses-palmer** [comment]
Can you provide a traceback? This could probably be done by running your script on a different VT by pressing <kbd>ctrl+alt+[1-9]</kbd> and then launching `DISPLAY=:0 python3 myscript.py 2>&1 | tee log`.

**Comment by katarina-ragna** [comment]
Hi!
I'm experiencing the same problem - system crash and log-out when suppress=True.
My system: Ubuntu 18.04. LTS, Python 3.6.9. And from `ps -e | grep X`
I get:
`10459 tty1     00:00:00 Xwayland
10640 tty2     00:01:14 Xorg`
Setting $DISPLAY as recommended [in the platform limitations](https://pynput.readthedocs.io/en/latest/limitations.html#linux) doesn't help.

The trouble is, I really would like to use pynput in my application, and the suppress=True option would improve usability a lot. However, without this option pynput is running without any problems (apart from the keyboard events being sent to the command line after my program is finished - which is exactly the behavior I'd like to suppress).

It would be great if you could specify the consequences that this Wayland issue mentioned in other posts has for the usage of pynput: Should pynput not be used in Wayland environments at all? How come I'm not experiencing the error when suppress=False? Is there a robust workaround to the crash problem?

Any help will be very much appreciated!

**Comment by katarina-ragna** [comment]
(By the way, the first few program executions are usually fine here, but after 3-4 runs the system reliably crashes -- this is confusing me a lot.)

**Comment by reapler** [comment]
i got the same issue on kubuntu 20.04 however with the mouse listener. I tried to pinpoint the problem, but it seems it happens too randomly on my side.
Later i tried to re-implement it from various sources and tinkered a working [solution](https://gist.github.com/reapler/4b04c413aec374176b12bb02fbe6a18e) for my case.
But prior i have also tested pynput and it seems it doesn't cleanup the record context if program is quit by user, signal or exception iirc.
Maybe the ungraceful exit could be a cause for this issue? dunno

**Comment by ochen1** [comment]
The keyboard listener with suppress=True crashed me too.

My code was extremely simple:

```py
from pynput import keyboard
import sys


def on_press(key):
    if hasattr(key, 'char'):
        print(key.char)
    sys.exit()


if __name__ == "__main__":
    with keyboard.Listener(on_press=on_press, suppress=True) as listener:
        listener.join()
```

I am using the `i3` window manager on KDE Neon.

**Comment by reapler** [comment]
I can confirm @ochen1' example. After the second execution it will logouts the user.
I have managed to fix the problem on kde neon 20.04. I believe it should also working on similar distros.
https://github.com/reapler/pynput/commit/5a1bbcf33519e9e5b6593e4be8949e1cdb6756de

However the display instance won't cleanup if the process is exit by user or signal. This may corrupt the desktop environment.
I think it is important to do a proper cleanup, doesn't matter under what circumstances since it is somewhat unmanaged(?) and could lead to the implied issues.

**Comment by moses-palmer** [comment]
The issues with a suppressing listener are presumed to be fixed on the *master* branch, so I will close this issue. If the problems persist, please reopen.
